File: backend/src/main/java/com/bobcarrental/service/AccountService.java
package com.bobcarrental.service;

import com.bobcarrental.dto.request.AccountRequest;
import com.bobcarrental.dto.response.AccountResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.math.BigDecimal;
import java.util.List;

/**
 * Service interface for account operations
 */
public interface AccountService {

    /**
     * Get all accounts with pagination
     */
    Page<AccountResponse> getAllAccounts(Pageable pageable);

    /**
     * Get account by ID
     */
    AccountResponse getAccountById(Long id);

    /**
     * Get account by code
     */
    AccountResponse getAccountByCode(String code);

    /**
     * Get accounts by type
     */
    List<AccountResponse> getAccountsByType(String accountType);

    /**
     * Get active accounts
     */
    List<AccountResponse> getActiveAccounts();

    /**
     * Get parent accounts (accounts without parent)
     */
    List<AccountResponse> getParentAccounts();

    /**
     * Get sub-accounts for a parent account
     */
    List<AccountResponse> getSubAccounts(Long parentAccountId);

    /**
     * Search accounts by name or code
     */
    Page<AccountResponse> searchAccounts(String query, Pageable pageable);

    /**
     * Create new account
     */
    AccountResponse createAccount(AccountRequest request);

    /**
     * Update existing account
     */
    AccountResponse updateAccount(Long id, AccountRequest request);

    /**
     * Update account balance
     */
    AccountResponse updateAccountBalance(Long id, BigDecimal amount, String transactionType);

    /**
     * Activate account
     */
    AccountResponse activateAccount(Long id);

    /**
     * Deactivate account
     */
    AccountResponse deactivateAccount(Long id);

    /**
     * Delete account
     */
    void deleteAccount(Long id);

    /**
     * Check if account code exists
     */
    boolean existsByCode(String code);
}

// Made with Bob

File: frontend/src/app/core/services/account.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, map } from 'rxjs';
import { environment } from '../../../environments/environment';
import { AccountRequest, AccountResponse, ApiResponse } from '../../models';

interface PageResponse<T> {
  content: T[];
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
  first: boolean;
  last: boolean;
  hasPrevious: boolean;
  hasNext: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class AccountService {
  private readonly API_URL = `${environment.apiUrl}/accounts`;

  constructor(private http: HttpClient) {}

  getAllAccounts(): Observable<AccountResponse[]> {
    return this.http.get<ApiResponse<PageResponse<AccountResponse>>>(this.API_URL).pipe(
      map(response => response.data.content)
    );
  }

  getAccountById(id: number): Observable<AccountResponse> {
    return this.http.get<ApiResponse<AccountResponse>>(`${this.API_URL}/${id}`).pipe(
      map(response => response.data)
    );
  }

  createAccount(request: AccountRequest): Observable<AccountResponse> {
    return this.http.post<ApiResponse<AccountResponse>>(this.API_URL, request).pipe(
      map(response => response.data)
    );
  }

  updateAccount(id: number, request: AccountRequest): Observable<AccountResponse> {
    return this.http.put<ApiResponse<AccountResponse>>(`${this.API_URL}/${id}`, request).pipe(
      map(response => response.data)
    );
  }

  deleteAccount(id: number): Observable<void> {
    return this.http.delete<ApiResponse<void>>(`${this.API_URL}/${id}`).pipe(
      map(() => undefined)
    );
  }

  activateAccount(id: number): Observable<AccountResponse> {
    return this.http.patch<ApiResponse<AccountResponse>>(`${this.API_URL}/${id}/activate`, {}).pipe(
      map(response => response.data)
    );
  }

  deactivateAccount(id: number): Observable<AccountResponse> {
    return this.http.patch<ApiResponse<AccountResponse>>(`${this.API_URL}/${id}/deactivate`, {}).pipe(
      map(response => response.data)
    );
  }
}

// Made with Bob